ZEN-1810

sparksam opened the pull request:
@philschatz Please review for me.

sparksam committed:
ZEN-1810
Created backend processes to loop through Github books and locate the corresponding CNXML Files.
Load Environment variables.
Added a caching system for quick responses to frontend avoid scrapping on every query.

sparksam committed:
ZEN-1810
WIP: Code currently really messy.
TODO: Need to write a more reusable and clean code
Fix CNXML redirection after search
Comment code.

sparksam committed:
ZEN-1810
WIP: refactored the seach process for both legacy and github books. Still need to figure out how to improve the cnxml search for Github hosted books.
Fix CNXML redirection after search

sparksam committed:
ZEN-1810
WIP: Removed the python files created earlier.

Commit:
Added Bootstrap 4 to update the look&feel
Updated the CNXML links to redirect to Github instead of downloading the files.

philschatz commented on app.js:
```diff
@@ -1,458 +1,573 @@
+                    book.versions.forEach(v => commit_shas.push(v.commit_sha.substring(0, 7)))
+                    for (let pipeline of pipelines) {
+                         for (let commitSha of commit_shas) {
+                              bookUrl = `${s3RootUrl}/${pipeline}/contents/${bookUuid}@${commitSha}.json`
```
It seems that the S3 URL is fetched even if the user has chosen to search CNXML files. If the S3 required what do you think of only using S3 when XHTML is selected?

![image](https://user-images.githubusercontent.com/253202/152827515-b826a016-7b81-4135-b73e-33108b126811.png)

sparksam replied:
I try to get the title of the book via the book json file. and I can only find the json on S3. any other idea how to retrieve it?

philschatz commented on app.js:
```diff
@@ -1,458 +1,573 @@
+               let collectionXMLLinks
+               if (!legacyBook) {
+                    //Download the book.xml from Github
+                    bookXML = await fetchWithBackoff(`${openstaxGithubURL}/${book.repository_name}/META-INF/books.xml`, false)
```
When GH Pages is not enabled, this fetch results in a 404 but gives a generic error message. Could this tell the user that one reason for the 404 message could be that the repository name in the ABL has a typo or another could be that the repo does not have GH-Pages enabled

sparksam replied:
Really good idea. Will implement and submit a PR

philschatz commented on app.js:
```diff
@@ -1,458 +1,573 @@
+                         try {
+                              const i = pageRefs.indexOf(pageRef)
+                              
+                              // Check this in the loop so that users can update the value while this is running
```
This check seems to be missing in the `if (!legacy)` branch above.

sparksam replied:
Nice catch, I 've put `if(legacy)` conditions inside the else to avoid  duplicate codes.

philschatz commented on app.js:
```diff
@@ -1,458 +1,573 @@
+                         if (pipelines.indexOf(v.min_code_version) === -1) pipelines.push(v.min_code_version)
+                         v.commit_metadata.books.forEach(book => {
+                              bookIdAndSlugs.push({...bookContainer, ...book})
+                              window.localStorage.setItem(book.uuid, JSON.stringify(bookContainer))
```
It seems the code always searches all books in a bundle (cool!) If the book-selection list contained just the bundle entry instead of the specific book in the bundle, would `localStorage` to store book JSON still be necessary?

philschatz reviewed (approved):
I just had a few suggestions but am excited to get this merged!

Also, all books that are in the ABL should now have a `https://github.io/.../books.xml` file that resolves

philschatz commented on app.js:
```diff
@@ -1,458 +1,573 @@
+               analyzeFn('BOOK:START', bookJson, bookUuid)
+               
+               
+               bookTitleEl.textContent = bookJson.title
```
If a valid codeversion cannot be found `bookJson` is null.
```suggestion
               if (bookJson === null) {
                   alert(`Could not find a baked version of ${bookUuid}@${commitSha}. Tried using these codeversions: ${JSON.stringify(pipelines)}`)
                   return
               }
               bookTitleEl.textContent = bookJson.title
```

sparksam committed:
Update app.js

If a valid codeversion cannot be found bookJson is null.

Co-authored-by: Philip Schatz <253202+philschatz@users.noreply.github.com>

Commit:
-Improved the exception handling
-Added an error log for more investigation for links that and errors.
-The search continues even some exceptions.
